lua/packer/neorocks.lua

```diff
@@ -0,0 +1,348 @@
+package.loaded['packer.util'] = nil
+package.loaded['packer.neorocks'] = nil
+package.loaded['packer.window'] = nil
+
+local a      = require('packer.async')
+local jobs   = require('packer.jobs')
+local log    = require('packer.log')
+local result = require('packer.result')
+local util   = require('packer.util')
+local window = require('packer.window')
+
+local await = a.wait
+local async = a.sync
+
+local neorocks = {}
+
+---@return table Of the form: {
+---     lua: Lua Version (5.1, 5.2, etc.)
+---     jit: Jit Version (2.1.0-beta3, or nil)
+---     dir: Directory of hererocks installation
+--- }
+neorocks._lua_version = (function()
+  if jit then
+    return {
+      lua = string.gsub(_VERSION, "Lua ", ""),
+      jit = string.gsub(jit.version, "LuaJIT ", ""),
+      dir = string.gsub(jit.version, "LuaJIT ", "")
+    }
+  end
+
+  error("NEOROCKS: Unsupported Lua Versions", _VERSION)
+end)()
+
+neorocks._base_path = util.join_paths(vim.fn.stdpath('cache'), 'packer_hererocks')
+neorocks._hererocks_file = util.join_paths(vim.fn.stdpath('cache'), 'hererocks.py')
+neorocks._hererocks_install_location = util.join_paths(neorocks._base_path, neorocks._lua_version.dir)
+neorocks._is_setup = vim.fn.isdirectory(util.join_paths(neorocks._hererocks_install_location, "lib")) > 0
+
+--- Determine if the hererocks environment is usable for nvim
+---
+---@return boolean True if the hererocks environment is already set up
+neorocks.is_setup = function()
+  return neorocks._is_setup
+end
+
+neorocks.get_hererocks = async(function()
+  local url_loc = 'https://raw.githubusercontent.com/luarocks/hererocks/latest/hererocks.py'
+
+  local cmd
+  if vim.fn.executable('curl') > 0 then
+    cmd = string.format(
+      'curl %s -o %s',
+      url_loc,
+      util.absolute_path(neorocks._hererocks_file)
+    )
+  elseif vim.fn.executable('wget') > 0 then
+    cmd = string.format(
+      'wget %s -O %s --verbose',
+      url_loc,
+      util.absolute_path(neorocks._hererocks_file)
+    )
+  else
+    error('"curl" or "wget" is required to install hererocks')
+  end
+
+  local r = result.ok(true)
+  return r:and_then(
+      await, jobs.run(cmd, {})
+    ):map_err(
+      function(err)
+        return {
+          msg = "Failed to get hererocks",
+          data = err
+        }
+      end
+    ):map_ok(
+      function(ok)
+        return {
+          status = ok,
+          -- TODO: Figure out what we should be doing with the output
```

Comment:
The output is usually collected on the `results` table (for installs and updates), for use in the operation status display. Here, we may not care about the output at all. We could probably just drop this entire `map_ok` call.